src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,203 @@
+ * <p></p>
+ * {@link uk.gov.hmcts.reform.bulkscanprocessor.tasks.BlobProcessorTask}
+ * <p></p>
+ * All the validations have been accomplished and {@link Envelope} saved in DB.
+ * Therefore no need to go through the same set of validation steps/rules and just re-extract zip file and upload.
+ * In case some {@link StorageException} or {@link IOException} experienced during zip extraction
+ * the envelope state will not be changed and left for a retry on the next run.
+ * {@link Event#DOC_UPLOAD_FAILURE} can be treated as stuck envelope and be re-uploaded by same service here.
+ * There is an upload retry counter/limit which may be used as well
+ * Review ^ last sentence and, perhaps, incorporate such possibility here.
+ */
+@Service
+public class UploadEnvelopeDocumentsService {
+
+    private static final Logger log = getLogger(UploadEnvelopeDocumentsService.class);
+
+    private final long coolDownPeriod;
+    private final EnvelopeRepository envelopeRepository;
+    private final BlobManager blobManager;
+    private final ZipFileProcessor zipFileProcessor;
+    private final ProcessEventRepository eventRepository;
+    private final DocumentProcessor documentProcessor;
+    private final EnvelopeProcessor envelopeProcessor;
+
+    public UploadEnvelopeDocumentsService(
+        // only process envelopes after cool down period
+        // can be removed once uploading feature is removed from main job
+        @Value("${scheduling.task.upload-documents.cool-down-minutes}") long coolDownPeriod,
+        EnvelopeRepository envelopeRepository,
+        BlobManager blobManager,
+        ZipFileProcessor zipFileProcessor,
+        ProcessEventRepository eventRepository,
+        DocumentProcessor documentProcessor,
+        EnvelopeProcessor envelopeProcessor
+    ) {
+        this.coolDownPeriod = coolDownPeriod;
+        this.envelopeRepository = envelopeRepository;
+        this.blobManager = blobManager;
+        this.zipFileProcessor = zipFileProcessor;
+        this.eventRepository = eventRepository;
+        this.documentProcessor = documentProcessor;
+        this.envelopeProcessor = envelopeProcessor;
+    }
+
+    public void processEnvelopes() {
+        envelopeRepository
+            .findByStatus(CREATED)
+            .stream()
+            // can be moved to query instead. but it won't be needed after upload is removed from main task
+            .filter(envelope -> envelope.getCreatedAt().isBefore(now().minus(coolDownPeriod, MINUTES)))
+            .collect(groupingBy(Envelope::getContainer))
+            .forEach(this::processByContainer);
+    }
+
+    private void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
+
+        try {
+            CloudBlobContainer blobContainer = blobManager.getContainer(containerName);
+
+            envelopes.forEach(envelope ->
+                getBlobInputStream(
```

Comment:
When processing an envelope, a lease must be acquired on it, so that multiple processes won't upload PDFs for it.

Reply:
ah yes, because it's a different job now